qa/1810

```diff
@@ -0,0 +1,65 @@
+#!/bin/sh
+# PCP QA Test No. 1810
+# Exercise the BPF PMDA CO-RE oomkill module - install, remove and values.
+#
+# Copyright (c) 2022 Sohaib Mohamed.
+#
+
+seq=`basename $0`
+echo "QA output created by $seq"
+
+. ./common.bpf
+
+_pmdabpf_check
+_pmdabpf_require_kernel_version 5 0
+_pmdabpf_require_libbpf_version 0 7
+
+status=1       # failure is the default!
+signal=$PCP_BINADM_DIR/pmsignal
+$sudo rm -rf $tmp.* $seq.full
+
+_cleanup()
+{
+    _pmdabpf_cleanup
+}
+
+_prepare_pmda bpf
+trap "_cleanup; exit \$status" 0 1 2 3 15
+_stop_auto_restart pmcd
+
+# real QA test starts here
+cat <<EOF | _pmdabpf_install
+# Installed by PCP QA test $seq on `date`
+[oomkill.so]
+enabled=true
+EOF
+_pmdabpf_wait_for_metric
+
+# Generate system activity for the CO-RE oomkill module
+perl -e 'while (1) { $a .= "A" x 1024; }'
```

Comment:
on a big machine, this could take a while, are there solutions like cgroup/ulimit that could achieve it quickly?

Reply:
@andreasgerstmayr any suggestions here, please? 
My reference: https://github.com/iovisor/bcc/blob/0e29f7032897f2ebdd4c617910aa8ade74bc4b23/tools/oomkill_example.txt#L36

Reply:
Good question. I'm not sure about the exact algorithm the oomkiller uses when deciding which process to kill, I think it's also possible that it kills another, unrelated process (which we want to avoid). I think it's a bit risky to trigger the oomkiller in a QA test.

What do you think @natoscott @kmcdonell, should we try to trigger the oom-killer as part of a QA test?

Reply:
Here's a cgroup example:

Single concat, runs inside 10MB
```
jkoch@jkoch:/sys/fs/cgroup/user.slice$ sudo systemd-run --scope -p MemoryLimit=10M perl -e '$a .= "A" x 1024'
Running scope as unit: run-r43d2f5e56b8847fea964d1d1a0bc19ec.scope
[completes]
```

With a loop, is terminated:
```
jkoch@jkoch:/sys/fs/cgroup/user.slice$ sudo systemd-run --scope -p MemoryLimit=10M perl -e 'while (1) { $a .= "A" x 1024 }'
Running scope as unit: run-r9ae1c624743547f58b4abdb773a73498.scope
Killed
```

Verifying:
```
jkoch@jkoch:/sys/fs/cgroup/user.slice$ sudo journalctl --since -5m | grep oom-kill 
Aug 10 09:44:29 jkoch kernel: perl invoked oom-killer: gfp_mask=0xcc0(GFP_KERNEL), order=0, oom_score_adj=0
Aug 10 09:44:29 jkoch kernel: oom-kill:constraint=CONSTRAINT_MEMCG,nodemask=(null),cpuset=run-rb41f35fa09a74858b93ba2537a1a955f.scope,mems_allowed=0,oom_memcg=/system.slice/run-rb41f35fa09a74858b93ba2537a1a955f.scope,task_memcg=/system.slice/run-rb41f35fa09a74858b93ba2537a1a955f.scope,task=perl,pid=100835,uid=0
```

Reply:
@andreasgerstmayr it does sound a little risky, and maybe also something that will be difficult to make reliable/deterministic across many kernel versions.  FWIW, there's also /proc/pid/oom_score that may help with focusing the oom-killer's attention, from memory its writable.